### PR TITLE
[Merged by Bors] - Directly extract joints into SkinnedMeshJoints

### PR DESCRIPTION
**Pull request by james7132**
# Objective
Following #4402, extract systems run on the render world instead of the main world, and allow retained state operations on it's resources. We're currently extracting to `ExtractedJoints` and then copying it twice during Prepare. Once into `SkinnedMeshJoints` and again into the actual GPU buffer.

This makes #4902 obsolete.

## Solution
Cut out the middle copy and directly extract joints into `SkinnedMeshJoints` and remove `ExtractedJoints` entirely.

This also removes the per-frame allocation that is being made to send `ExtractedJoints` into the render world.

## Performance
On my local machine, this halves the time for `prepare_skinned _meshes` on `many_foxes` (195.75us -> 93.93us on average).

image

---

## Changelog
Added: `BufferVec::truncate`
Added: `BufferVec::extend`
Changed: `SkinnedMeshJoints::build` now takes a `&mut BufferVec` instead of a `&mut Vec` as a parameter.
Removed: `ExtractedJoints`.

## Migration Guide
`ExtractedJoints` has been removed. Read the bound bones from `SkinnedMeshJoints` instead.

### CODE REVIEW

**Comment by james7132** [comment]
Curiously this also improves `extract_skinned_meshes` (417.23us -> 252.3us per frame, on my local machine).

image

File: crates/bevy_pbr/src/render/mesh.rs [hunk 1]
```diff
@@ -236,21 +232,19 @@ pub fn extract_skinned_meshes(
         }
         // PERF: This can be expensive, can we move this to prepare?
```
Comment (IceSentry) [inline crates/bevy_pbr/src/render/mesh.rs:233]: Is this comment still relevant/would it be possible to move it to prepare?

File: crates/bevy_pbr/src/render/mesh.rs [hunk 2]
```diff
@@ -188,19 +183,20 @@ impl SkinnedMeshJoints {
...
-                return None;
-            }
+        let target = start + skin.joints.len().min(MAX_JOINTS);
+        buffer.extend(
+            joints
+                .iter_many(&skin.joints)
+                .zip(inverse_bindposes.iter())
+                .map(|(joint, bindpose)| joint.affine() * *bindpose),
+        );
+        if buffer.len() != target {
```
Comment (IceSentry) [inline crates/bevy_pbr/src/render/mesh.rs:197]: I'm not entirely when this condition would be true. Does it mean that the query didn't contain all the joints?

**Review by IceSentry — APPROVED** [review]
Overall it looks good to me and many_foxes did work correctly as expected. Would be nice to have a bit more comments explaining what's going on though.

**Review by mockersf — APPROVED** [review]

**Comment by james7132** [comment]
bors retry

**Pull request closed without merging**